### PR TITLE
[Validator] - Added Timezone Constraint Validator

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Q | A |
| --- | --- |
| Bug fix? | [no] |
| New feature? | [yes] |
| BC breaks? | [no] |
| Deprecations? | [no] |
| Tests pass? | [yes] |
| Fixed tickets | [] |
| License | MIT |
| Doc PR | yes (symfony/symfony-docs#3907) |

comments: was previously, https://github.com/symfony/symfony/pull/11010. Re-created because I've deleted my old fork.

ping @webmozart

### CODE REVIEW

**Comment by xabbuh** [comment]
The test class looks like it's just a copy of the constraint validator class.

**Comment by mickaelandrieu** [comment]
@xabbuh already fixed, but well catched :dancer: 

Should be good now :)

File: src/Symfony/Component/Validator/Constraints/Timezone.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ * @Target({"PROPERTY", "METHOD", "ANNOTATION"})
+ *
+ * @author Mickaël Andrieu <mickael.andrieu@sensiolabs.com>
```
Comment (hhamon) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: Are you sure?

File: src/Symfony/Component/Validator/Constraints/Timezone.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @api
```
Comment (hhamon) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: I don't think you're allowed to tag with `@api` yourself.

File: src/Symfony/Component/Validator/Tests/Constraints/TimezoneValidatorTest.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
+<?php
+namespace Symfony\Component\Validator\Tests\Constraints;
```
Comment (hhamon) [inline src/Symfony/Component/Validator/Tests/Constraints/TimezoneValidatorTest.php]: Missing carriage return before `namespace` keyword and license header section.

Comment (xabbuh) [inline src/Symfony/Component/Validator/Tests/Constraints/TimezoneValidatorTest.php]: and missing the license block

File: src/Symfony/Component/Validator/Constraints/Timezone.php (hunk 1 above)
Comment (mickaelandrieu) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: was from sensiolabs when I did this pull request.
Extracted from a project with project manager agreement ;)

File: src/Symfony/Component/Validator/Constraints/Timezone.php (hunk 2 above)
Comment (mickaelandrieu) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: I didn't know, do I need to remove it so ? /c @webmozart

**Comment by mickaelandrieu** [comment]
@hhamon @xabbuh thank for this final review.

ping @webmozart : now this is realy done i guess :cat:

File: src/Symfony/Component/Validator/Constraints/Timezone.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * @Annotation
+ * @Target({"PROPERTY", "METHOD", "ANNOTATION"})
+ *
+ * @author Mickaël Andrieu <andrieu.travail@gmail.com>
+ *
+ */
+class Timezone extends Constraint
+{
+    public $message = 'The value "{{ value }}" is not a valid timezone';
```
Comment (jakzal) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: This message should be also added to [the translations file](https://github.com/symfony/symfony/blob/ad2f906daf8543a7034ef2f0eebd4531084f5fbe/src/Symfony/Component/Validator/Resources/translations/validators.en.xlf#L309) (at least the English one).

Comment (jakzal) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: Also, how about changing the message to `"{{ value }}" is not a valid timezone`? 

`The value "Foo/Bar" is not a valid timezone` doesn't sound right to me.

Comment (mickaelandrieu) [inline src/Symfony/Component/Validator/Constraints/Timezone.php]: Hi @jakzal, you're right, I've updated.

**Comment by stof** [comment]
Please also add test cases against the other validation APIs

**Comment by mickaelandrieu** [comment]
@stof need help here as I don't know what I need to do, did you have an example ?

File: src/Symfony/Component/Validator/Resources/translations/validators.en.xlf [hunk 5]
```diff
@@ -306,6 +306,10 @@
                 <source>The host could not be resolved.</source>
                 <target>The host could not be resolved.</target>
             </trans-unit>
+            <trans-unit id="80">
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Resources/translations/validators.en.xlf]: The id will need to become `81`.

**Comment by xabbuh** [comment]
@mickaelandrieu You can have a look at the `Legacy*ValidatorLegacyApiTest` classes.

**Comment by mickaelandrieu** [comment]
After 12 months I think ultimately this constraint does not need to be part of the core . If it had been, it would have been merged long time ago  (was validated for 2.4, then 2.5, then 2.6, now 2.7 or 3.0 ?).

I'm sorry for all who have reviewed this one, but I don't want to "waste" time on this (and I don't want others waste time on this too), because this is not so useful and it's better at this point to just close this.

**Pull request closed without merging**